sw/device/tests/dif/dif_gpio_test.cc

```diff
@@ -0,0 +1,458 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+#include <limits>
+
+extern "C" {
+#include "sw/device/lib/dif/dif_gpio.h"
+
+#include "gpio_regs.h"  // Generated
+}  // extern "C"
+
+// Base class for the test fixtures in this file.
+class DifGpioTest : public testing::Test, public mock_mmio::MmioTest {
+ protected:
+  static constexpr uint32_t kAllOnes = std::numeric_limits<uint32_t>::max();
+};
+
+// Init tests
+class InitTest : public DifGpioTest {};
+
+TEST_F(InitTest, NullArgs) {
+  dif_gpio_config_t config{.base_addr = dev().region()};
+  dif_gpio_t out_arg_gpio;
+
+  EXPECT_EQ(dif_gpio_init(nullptr, &out_arg_gpio), kDifGpioBadArg);
+  EXPECT_EQ(dif_gpio_init(&config, nullptr), kDifGpioBadArg);
+  EXPECT_EQ(dif_gpio_init(nullptr, nullptr), kDifGpioBadArg);
+}
+
+TEST_F(InitTest, Init) {
+  // Note: `dif_gpio_init` does not write to `GPIO_MASKED_OE_*` and
+  // `GPIO_MASKED_OUT_*` since it directly resets `GPIO_DIRECT_OE` and
+  // `GPIO_DIRECT_OUT`.
+  EXPECT_WRITE32(GPIO_INTR_ENABLE_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_DIRECT_OE_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_DIRECT_OUT_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_FALLING_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_LVLHIGH_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_STATE_REG_OFFSET, kAllOnes);
+
+  dif_gpio_config_t config{.base_addr = dev().region()};
+  dif_gpio_t gpio;
+  EXPECT_EQ(dif_gpio_init(&config, &gpio), kDifGpioOk);
+}
+
+// Base class for the rest of the tests in this file, provides a
+// `dif_gpio_t` instance.
+class DifGpioTestInitialized : public DifGpioTest {
+ protected:
+  uint32_t AllZerosExcept(uint32_t index) { return 1 << index; }
```

Comment:
Similarly, I feel these should be global free functions, since they do not use the test fixture for anything.